Add ansor blog post (#23)
apache/tvm-site: _posts/2021-03-01-intro-auto-scheduler.md

---
layout: post
title: Introducing TVM Auto-scheduler (a.k.a. Ansor)

> date: 2021-03-01

Reviewer: please remember to change the time and file title to merge date after merge

author: Lianmin Zheng, Chengfan Jia, Minmin Sun, Zhao Wu, Cody Hao Yu
---

Optimizing the execution speed of deep neural networks is extremely hard with the growing
model size, operator diversity, and hardware heterogeneity.
From a computational perspective, deep neural networks are just layers and layers of tensor computations.
These tensor computations, such as matmul and conv2d, can be easily described by mathematical expressions.
However, providing high-performance implementations for them on modern hardware can be very challenging.
We have to apply various low-level optimizations and utilize special hardware intrinsics to achieve high performance.
It takes huge engineering effort to build linear algebra and neural network acceleration libraries like CuBLAS, CuDNN, oneMKL, and oneDNN.

Our life will be much easier if we can just write mathematical expressions and have something
magically turn them into efficient code implementations.

> Three years ago, compiler TVM and its search module AutoTVM were built as the first step towards this goal.

Reviewer 2: compiler TVM -> Deep learning compiler TVM

AutoTVM employs a template-based search algorithm to find efficient implementations for a given tensor computation.
However, it is a template-based approach, so it still requires domain experts to implement a non-trivial manual template
for every operator on every platform.
Today, there are more than 15k lines of code for these templates in the TVM code repository.

> Besides being very hard to develop, these templates often have limited search spaces,

Reviewer 3: People might think AutoTVM search space is already very large, so it would be clearer to say it is inefficient.
```suggestion
Besides being very hard to develop, these templates often have inefficient and limited search spaces,
```

making them unable to achieve optimal performance.

> To address the limitations of AutoTVM, we started project Ansor to build a fully automated auto-scheduler for code generation.

Reviewer 2: I prefer replacing `code generation` into `generating high performance tensor programs`. As code generation here maybe indicate we do `CodeGenLLVM / CodeGenOpenCL` work, but those work have been done by TVM before. We focus on generating high performance tensor program as our paper shows.

> Ansor auto-scheduler only takes tensor expressions as input and generates high-performance code without manual templates.

Reviewer 2: Ansor auto-scheduler -> Ansor, I think just one term is enough.

In step 2, the developer has to write a schedule template, which typically consists of 20-100 lines of tricky DSL code.
This part requires domain expertise of both the target hardware architecture and operator semantics, so it is difficult.
The last step, step 3, is automated by a search algorithm.

In auto-scheduler, we eliminate the most difficult step 2 by automatic search space construction and accelerate step 3 with a better search algorithm.
By doing automatic search space construction, we not only eliminate huge manual effort, 
but also enabling the exploration of much more optimization combinations.
This automation does not come free, because we still need to design rules to generate the search space.
However, these rules are very general. They are based on static analysis of the tensor expressions.
We only need to design a few general rules once and can apply them to almost all tensor computations in deep learning.

## Search Process
{:center: style="text-align: center"}
![image](/images/intro-auto-scheduler/search_overview.png){: width="40%"}
{:center}
<center> Figure 1. Search Process Overview  </center> <p></p>

> Figure 1. shows the search process of auto-scheduler when optiming a whole neural network.

Reviewer 3: ```suggestion
Figure 1 shows the search process of auto-scheduler when optimizing a whole neural network.
```

> Figure 1. shows the search process of auto-scheduler when optiming a whole neural network.

Reviewer 2: optiming -> optimizing

The system takes deep learning models as input.
It then partitions the big model into small subgraphs with Relay's operator fusion pass.
A task scheduler is utilized to allocate the time resource for optimizing many subgraphs.
At each iteration, it picks a subgraph that has the most potential to increase the end-to-end performance.

> For this subgraph, we analysis its tensor expression and generates several sketches for it.

Reviewer 3: ```suggestion
For this subgraph, we analyze its tensor expression and generate several sketches for it.
```

Then we run evolutionary search with a learned cost model to get a batch of optimized programs.
The optimized programs are sent to actual hardware for measurements.
When the measurements are finished, the profiling results are used as feedback to update all components of the system.
This process is repeated iteratively until the optimization converges or we run out of time budget.
More technical details can be found in our paper [3] and our code.

> The auto-scheduler reuses the existing computation definitions in TOPI but does not use any schedule template.

Reviewer 3: ```suggestion
It is worth notiing that since the auto-scheduler generates schedules from scratch, it only reuses the existing computation definitions in TOPI but not schedule templates.
```

# Benchmark Results
In this section, we benchmark the performance of AutoTVM and Auto-scheduler.
The CPU benchmark is done on an AWS c5.9xlarge, which is equipped with an Intel 18-core skylake 8124-m CPU. 
The GPU benchmark is done on an AWS g4dn.4xlarge, which is equipped with an NVIDIA T4 GPU.
All benchmark code, raw data, tuning logs can be found in this repo [2].

## Performance of the generated code
We benchmark the fp32 single-batch inference latency on three networks.

> Figure 2 shows the relative speedup auto-scheduler can get against AutoTVM.

Reviewer 3: ```suggestion
Figure 2 shows the relative speedup auto-scheduler against to AutoTVM.
```

We can see auto-scheduler outperforms AutoTVM in all cases with 1.02x to 8.95x speedup.

> This is because auto-scheduler explores a larger search space. Auto-scheduler can find more efficient combinations
> of optimizations that are missed by manual templates.

Reviewer 3: ```suggestion
This is because auto-scheduler explores a larger search space, which covers more efficient combinations
of optimizations that are missed in TOPI manual templates.
```

The BERT-base@GPU is an extreme case where the manual templates are very badly designed.

> The manual template for dense layers does not perform well for the shapes in BERT model.

Reviewer 3: ```suggestion
In other words, the manual template for dense layers does not perform well for the shapes in BERT model.
Instead of leveraging an expert to manually optimize the dense schedule template, we can now simply push a button to get much better performance with the auto-scheduler.
```

{:center: style="text-align: center"}
![image](/images/intro-auto-scheduler/code_perf.png){: width="85%"}
{:center}
<center> Figure 2. Code Performance Comparision (Higher is better) </center> <p></p>

## Search Time
As we know, the search-based approaches can be very time-consuming, so we also care about the search time.
It typically takes several hours to let the search converge for a single neural network.
Figure 3 compares the search time of AutoTVM and auto-scheduler.
Auto-scheduler requires much less time to converge in most cases, despite its larger search space.

> This is because of auto-scheduler has a better cost model and task scheduler.

Reviewer 3: ```suggestion
This is mainly because auto-scheduler has a better cost model and task scheduler.
```

{:center: style="text-align: center"}
![image](/images/intro-auto-scheduler/search_time.png){: width="85%"}
{:center}
<center> Figure 3. Search Time Comparision (Lower is better) </center> <p></p>

## More Results
The repo above serves as an internal benchmark tool for TVM, so it only compares the latest AutoTVM and AutoScheduler.
You can find results for more libraries and backends in our paper [3].
Recently, this blog post [4] also tried auto-scheduler on an Apple M1 chip and got some good results.

# Conclusion
We build TVM auto-scheduler, a system that automatically generates high-performance code for tensor expressions.

> Compared with the predecessor AutoTVM, auto-scheduler does not require manual templates.
> Besides, auto-scheduler generates better code with less search time.

Reviewer 3: ```suggestion
Compared with the predecessor AutoTVM, auto-scheduler does not require manual templates, but
auto-scheduler is capable of generating schedules with better performance in a shorter time.
```
